src/Components/Endpoints/src/TempData/TempData.cs

```diff
@@ -0,0 +1,235 @@
+    }
+
+    /// <summary>
+    /// Loads data from a <paramref name="data"/> into the TempData dictionary.
+    /// </summary>
+    public void Load(IDictionary<string, object?> data)
+    {
+        _data.Clear();
+        _retainedKeys.Clear();
+        foreach (var kvp in data)
+        {
+            _data[kvp.Key] = kvp.Value;
+            _retainedKeys.Add(kvp.Key);
+        }
+    }
+
+    /// <inheritdoc/>
+    public void Clear()
+    {
+        EnsureLoaded();
+        _data.Clear();
+        _retainedKeys.Clear();
+    }
+
+    ICollection<string> IDictionary<string, object?>.Keys => _data.Keys;
+
+    ICollection<object?> IDictionary<string, object?>.Values => _data.Values;
+
+    int ICollection<KeyValuePair<string, object?>>.Count => _data.Count;
+
+    bool ICollection<KeyValuePair<string, object?>>.IsReadOnly => ((ICollection<KeyValuePair<string, object?>>)_data).IsReadOnly;
+
+    void IDictionary<string, object?>.Add(string key, object? value)
+    {
+        this[key] = value;
+    }
+
+    bool IDictionary<string, object?>.TryGetValue(string key, out object? value)
+    {
+        value = Get(key);
+        return ContainsKey(key);
+    }
+
+    void ICollection<KeyValuePair<string, object?>>.Add(KeyValuePair<string, object?> item)
+    {
+        ((IDictionary<string, object?>)this).Add(item.Key, item.Value);
+    }
+
+    bool ICollection<KeyValuePair<string, object?>>.Contains(KeyValuePair<string, object?> item)
+    {
+        return ContainsKey(item.Key) && Equals(Peek(item.Key), item.Value);
+    }
+
+    void ICollection<KeyValuePair<string, object?>>.CopyTo(KeyValuePair<string, object?>[] array, int arrayIndex)
+    {
+        EnsureLoaded();
+        ((ICollection<KeyValuePair<string, object?>>)_data).CopyTo(array, arrayIndex);
+    }
+
+    bool ICollection<KeyValuePair<string, object?>>.Remove(KeyValuePair<string, object?> item)
+    {
+        if (ContainsKey(item.Key) && Equals(Peek(item.Key), item.Value))
+        {
+            return Remove(item.Key);
+        }
+        return false;
+    }
+
+    IEnumerator<KeyValuePair<string, object?>> IEnumerable<KeyValuePair<string, object?>>.GetEnumerator()
+    {
+        EnsureLoaded();
+        return new TempDataEnumerator(this);
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        EnsureLoaded();
+        return new TempDataEnumerator(this);
+    }
+
+    class TempDataEnumerator : IEnumerator<KeyValuePair<string, object?>>
+    {
+        private readonly TempData _tempData;
+        private readonly IEnumerator<KeyValuePair<string, object?>> _innerEnumerator;
+
+        public TempDataEnumerator(TempData tempData)
+        {
+            _tempData = tempData;
+            _innerEnumerator = tempData._data.GetEnumerator();
+        }
+
+        public KeyValuePair<string, object?> Current
+        {
+            get
+            {
+                var kvp = _innerEnumerator.Current;
+                _tempData.Remove(kvp.Key);
```

Comment:
nit:
TempData implemets `IEnumerable`, so calling `foreach` on it is possible. It can cause
```
Potential InvalidOperationException - modifying collection during enumeration
```

E.g.
```
foreach (var item in tempData)
{
    Console.WriteLine(item.Key);  // Accessing Current triggers Remove
}
```

We should avoid it, e.g. by collecting keys for removal and do it after enumeration (use dispose method) or by creating a snapshot for enumeration (enumeration on a copy).